### PR TITLE
Add support for default configuration modes

### PR DESCRIPTION
**Pull request by alextwoods**
Adds support for the `default_mode` configuration value that standardizes SDK default configurations.

Add a new aws-defaults pseudo gem (similar to seahorse and sts/sso) in core which has:
* `DefaultsModeConfigResolver` - responsible for logic to resolve defaults from a defaults mode (as well as environment detection).
* `DefaultsModeConfiguration` module - documentation and config values code generated from the cross SDK values.

`default_mode` is added as an option by the `DefaultsMode` plugin.  Resolution of defaults is done through updates to the default blocks on the appropriate existing config options.

This change also adds support for the `ssl_timeout` config option for net http.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. To make sure we include your contribution in the release notes, please make sure to add description entry for your changes in the "unreleased changes" section of the `CHANGELOG.md` file (at corresponding gem). For the description entry, please make sure it lives in one line and starts with `Feature` or `Issue` in the correct format.

2. For generated code changes, please checkout below instructions first:
  https://github.com/aws/aws-sdk-ruby/blob/version-3/CONTRIBUTING.md

Thank you for your contribution!

### CODE REVIEW

**Comment by alextwoods** [comment]
Reminder: Confirm core version requirements before merge/release.

File: tasks/update-defaults-mode.rake [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  defaults_file = args[:defaults_file]
+  if defaults_file.nil? || defaults_file.empty?
+    raise ArgumentError, 'Missing required defaults_file, please provide one as a rake argument.'
+  end
+
+  puts "Loading defaults from: #{defaults_file}"
+  defaults = JSON.load(File.read(defaults_file))
+
+  default_mode_docs = ["<p>The following `:default_mode` values are supported: </p>"]
```
Comment (mullermp) [inline tasks/update-defaults-mode.rake]: Should these hardcoded strings live in the actual documentation, and maybe just code generate multiple dynamic sections like all of the modes and configuration maps?

File: gems/aws-sdk-s3/spec/client/iad_regional_endpoint_spec.rb [hunk 2]
```diff
@@ -63,6 +63,20 @@ module S3
             'https://s3.us-east-1.amazonaws.com/')
         end
 
+        it 'can be set through defaults mode' do
```
Comment (mullermp) [inline gems/aws-sdk-s3/spec/client/iad_regional_endpoint_spec.rb:66]: Did you update all of the plugin tests?

File: gems/aws-sdk-s3/lib/aws-sdk-s3/client.rb [hunk 3]
```diff
@@ -218,6 +220,11 @@ class Client < Seahorse::Client::Base
     #     Used only in `standard` and adaptive retry modes. Specifies whether to apply
     #     a clock skew correction and retry requests with skewed client clocks.
     #
+    #   @option options [String] :defaults_mode ("legacy")
+    #     TODO: See Aws::ConfigurationDefaults Documentation
```
Comment (mullermp) [inline gems/aws-sdk-s3/lib/aws-sdk-s3/client.rb:224]: You can link with `{Aws::ConfigurationDefaults}` https://gist.github.com/phansch/db18a595d2f5f1ef16646af72fe1fb0e#linking-objects-

File: gems/aws-sdk-dynamodb/lib/aws-sdk-dynamodb/client.rb [hunk 4]
```diff
@@ -187,6 +189,11 @@ class Client < Seahorse::Client::Base
     #     Used only in `standard` and adaptive retry modes. Specifies whether to apply
     #     a clock skew correction and retry requests with skewed client clocks.
     #
+    #   @option options [String] :defaults_mode ("legacy")
+    #     TODO: See Aws::ConfigurationDefaults Documentation
+    #
+    #   @option options [Aws::DefaultsModeConfigResolver] :defaults_mode_config_resolver
```
Comment (mullermp) [inline gems/aws-sdk-dynamodb/lib/aws-sdk-dynamodb/client.rb]: Why is this an allowable option? This seems api private

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/retry_errors.rb [hunk 5]
```diff
@@ -163,8 +163,14 @@ class RetryErrors < Seahorse::Client::Plugin
       option(:clock_skew) { Retries::ClockSkew.new }
 
       def self.resolve_retry_mode(cfg)
-        value = ENV['AWS_RETRY_MODE'] ||
+        default_mode_value =
+          if cfg.respond_to?(:defaults_mode_config_resolver)
+            cfg.defaults_mode_config_resolver.resolve(:retry_mode)
+          end
+
+          value = ENV['AWS_RETRY_MODE'] ||
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/retry_errors.rb:171]: formatting

**Review by mullermp — COMMENTED** [review]
Nice!

File: tasks/update-defaults-mode.rake (hunk 1 above)
Comment (alextwoods) [reply tasks/update-defaults-mode.rake]: Yeah - I think that makes sense.

File: gems/aws-sdk-s3/spec/client/iad_regional_endpoint_spec.rb (hunk 2 above)
Comment (alextwoods) [reply gems/aws-sdk-s3/spec/client/iad_regional_endpoint_spec.rb:66]: Yeah -  every option that has changed has a new test.

File: tasks/update-defaults-mode.rake (hunk 1 above)
Comment (alextwoods) [reply tasks/update-defaults-mode.rake]: NVM - this doesn't work as we need all of the links to be processed together by kramdown.

**Comment by ixti** [comment]
It's a bit unclear how to use default config mode.

**Comment by mullermp** [comment]
You can set the `defaults_mode` option to any of the keys under this `modes` key in this JSON: https://github.com/aws/aws-sdk-ruby/blob/9d893a53136025661f8f55b6d3739df64d8c69f0/gems/aws-sdk-core/lib/aws-defaults/default_configuration.rb#L122

It is just a canned way of overwriting some other config options. You may be better off just overwriting any options you want instead.

**Comment by ixti** [comment]
Thank you!

**Pull request merged**